File: apps/patter-voice-sdk/libraries/python/getpatter/observability/event_bus.py
"""Lightweight synchronous/async event bus for Patter pipeline events."""

from __future__ import annotations

import asyncio
import logging
from typing import Any, Callable, Literal

logger = logging.getLogger(__name__)

PatterEventType = Literal[
    "turn_started",
    "turn_ended",
    "eou_metrics",
    "interruption",
    "llm_metrics",
    "tts_metrics",
    "stt_metrics",
    "metrics_collected",
    "call_ended",
    # Fine-grained pipeline events (additive — existing callbacks remain).
    "transcript_partial",
    "transcript_final",
    "llm_chunk",
    "tts_chunk",
    "tool_call_started",
]


class EventBus:
    """Lightweight event emitter.

    Emits are fire-and-forget; handlers that raise are caught and logged so
    a misbehaving observer never disrupts the call pipeline.

    Usage::

        bus = EventBus()
        unsubscribe = bus.on("turn_ended", lambda payload: print(payload))
        bus.emit("turn_ended", {"turn_index": 0})
        unsubscribe()  # remove listener
    """

    def __init__(self) -> None:
        self._listeners: dict[str, list[Callable[[Any], Any]]] = {}

    def on(
        self,
        event: PatterEventType,
        cb: Callable[[Any], Any],
    ) -> Callable[[], None]:
        """Register *cb* for *event*.

        Returns a zero-argument callable that removes the listener when called.
        """
        self._listeners.setdefault(event, []).append(cb)

        def _unsubscribe() -> None:
            listeners = self._listeners.get(event)
            if listeners is not None:
                try:
                    listeners.remove(cb)
                except ValueError:
                    pass

        return _unsubscribe

    def emit(self, event: PatterEventType, payload: Any) -> None:
        """Fire *event* with *payload* to all registered listeners.

        Synchronous callbacks are called inline. Async callbacks are scheduled
        via :func:`asyncio.create_task` (requires a running event loop).
        Exceptions are caught and logged; they never propagate to the caller.
        """
        for cb in list(self._listeners.get(event, [])):
            try:
                result = cb(payload)
                if hasattr(result, "__await__"):
                    task = asyncio.create_task(result)

                    def _log_listener_exc(t: "asyncio.Task[object]") -> None:
                        if t.cancelled():
                            return
                        exc = t.exception()
                        if exc is not None:
                            logger.error(
                                "event_bus async listener raised", exc_info=exc
                            )

                    task.add_done_callback(_log_listener_exc)
            except Exception:
                logger.exception("event_bus listener raised")

File: apps/patter-voice-sdk/libraries/typescript/src/observability/event-bus.ts
/**
 * Lightweight in-process event bus for Patter call lifecycle events.
 *
 * Mirrors the Python ``PatterEventBus`` (libraries/python/getpatter/observability/event_bus.py).
 * Consumers subscribe with ``on()`` and receive typed payloads.  ``emit()`` is
 * synchronous but handles async listeners: rejections are surfaced via the
 * Patter logger rather than being swallowed or crashing the call.
 */

import { getLogger } from '../logger';

/** String tag identifying every event type the `EventBus` knows how to dispatch. */
export type PatterEventType =
  | 'turn_started'
  | 'turn_ended'
  | 'eou_metrics'
  | 'interruption'
  | 'llm_metrics'
  | 'tts_metrics'
  | 'stt_metrics'
  | 'metrics_collected'
  | 'call_ended'
  // Fine-grained pipeline events (additive — existing callbacks remain).
  | 'transcript_partial'
  | 'transcript_final'
  | 'llm_chunk'
  | 'tts_chunk'
  | 'tool_call_started'
  // Pause-and-resume (bargeInMode: 'pause_resume'): a pause that no
  // transcript confirmed resumed from the unheard tail. Payload:
  // ``{ resumedSentences: number }``. Mirrors the Python emit.
  | 'false_interruption';

type Listener<T = unknown> = (payload: T) => void | Promise<void>;

/** In-process pub/sub for Patter call-lifecycle events. */
export class EventBus {
  private readonly listeners = new Map<PatterEventType, Set<Listener>>();

  /**
   * Subscribe to an event type.  Returns an unsubscribe function.
   */
  on<T = unknown>(event: PatterEventType, cb: Listener<T>): () => void {
    let set = this.listeners.get(event);
    if (!set) {
      set = new Set();
      this.listeners.set(event, set);
    }
    set.add(cb as Listener);
    return () => set!.delete(cb as Listener);
  }

  /**
   * Emit an event synchronously.  Async listeners are fire-and-forget with
   * rejection logging so a badly-behaved observer never stalls the call path.
   */
  emit<T = unknown>(event: PatterEventType, payload: T): void {
    const set = this.listeners.get(event);
    if (!set) return;
    for (const cb of [...set]) {
      try {
        const res = cb(payload);
        if (res && typeof (res as Promise<unknown>).catch === 'function') {
          (res as Promise<unknown>).catch((e) =>
            getLogger().error(`[EventBus] listener for "${event}" rejected:`, e),
          );
        }
      } catch (e) {
        getLogger().error(`[EventBus] listener for "${event}" threw:`, e);
      }
    }
  }
}
